### DISCUSSION TITLE
Road Bridge

### DISCUSSION THREAD

**Discussion by IsotoCedex** [Universite-Gustave-Eiffel/NoiseModelling#585] [Q&A]
Hello everyone,
I am trying to model a scenario with a road bridge.
I have thought of raising the section of road by modifying the relative heights on the ground.
That is, instead of assigning 0.05, raise the section of road above the bridge, with the relative height above the ground, plus 0.05m.
Also, include a "floating building", using the "Zground and Zobject" elements to simulate the effect of the bridge under the road.
What is your opinion? Is there any other way to simulate the bridge?
Thanks so much!

**Comment by pierromond** [comment]
Hi Ignacio,

Did you try the solution you are saying, normally it should work !

Tell us if not, and also if yes ;)

**Comment by IsotoCedex** [comment]
Hi Pierre,
No, I haven't.
I have not tried this since I have verified that the software forces the source to 0.05m above the ground. Even if I put a floating building on it, the road would be under the building.
After asking that question on GitHub, and rereading the manual, I thought about it, and I think that solution is not viable.
I'm right?

**Comment by pierromond** [comment]
It looks that you are using Road_emission_from traffic, and you are right there is this line : https://github.com/Universite-Gustave-Eiffel/NoiseModelling/blob/209200e9ba7ae256f1e99a67859b0756806d682d/wps_scripts/src/main/groovy/org/noise_planet/noisemodelling/wps/NoiseModelling/Road_Emission_from_Traffic.groovy#L228

You can comment this line and in this way it will not force the height to 0.05... Do you know how to modify groovy scripts ?

**Comment by IsotoCedex** [comment]
Okay. I try.
I understand that I must modify the file with an editor, and then run the program.

**Comment by IsotoCedex** [comment]
**AUTHOR'S NOTE: THIS INTERVENTION IS COMPLETELY WRONG. I LEAVE IT FOR A BETTER UNDERSTANDING OF THE DISCUSSION. DO NOT ATTEMPT TO REPRODUCE WHAT IS INDICATED IN IT.**


Hello again,
I have already managed to get the rays to come out from the road over the river:

![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/306fcd39-a1e3-4cf8-b10e-62b795d63f7c)

I therefore believe that NM understands that there is a road elevated over the DEM.

I have tried generating the bridge. I think I'm misunderstanding the instructions. I have followed the instructions in the following link:
https://noisemodelling.readthedocs.io/en/4.x/Input_buildings.html

The generated "floating bridge-building" is described as follows. It rests on the DEM, at the beginig and end of the riverbank. It starts at -1m height from the DTM, and has a thickness of 1m. So, the surface of the bridge match the DEM surface (I think...)

![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/4d64fde1-4702-4480-bf5c-b4c571ead932)
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/1e030776-d4c3-4110-b0d7-10202dddebd2)

I think I'm not doing it right. The result is the same with bridge and without bridge.

What I get with NM is:
- Without the bridge
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/71070070-91c3-4eb1-9f13-1398f498d369)
- With the bridge
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/abdf3076-9231-43e0-9522-d46b5b145318)


With other software the result is like this:
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/9ad64aa4-6eb2-4b19-80c3-b038cb2a48d2)

Any idea?

Thanks so much!

**Comment by nicolas-f** [comment]
Hi,

You could use this method to change the DEM to fit with the bridge height:

https://noisemodelling.readthedocs.io/en/latest/Input_dem.html#dem-enrichment

**Comment by IsotoCedex** [comment]
Hi again,
After a lot of attempts, I think I got it.
@nicolas-f, you were right. The key is on the next tutorials:
https://noisemodelling.readthedocs.io/en/latest/Input_dem.html#dem-enrichment
https://noisemodelling.readthedocs.io/en/latest/Input_buildings.html

It is necessary to manage a good terrain and to ensure that the bridge vertices are well defined.

![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/6ff30727-b5a7-4170-82c2-6149e7c07cf0)
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/83663892-8b28-4b63-b397-da7b8d2778a8)

Results of the first successful attempt are different than expected, but reasonable, in my opinion. Probably in further tests I could get more similar results.

![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/6ec21e80-d457-428b-84e0-779e0249e6ff)


Conclusion: it is possible to model bridges with noise modeling. I even think it would be possible to model bridges with acoustic screens. Vertices have to be over terrain. No floating secreens can be modeled, as far as I know.
It's great to have such a powerful free software.
Good job!! Thank you very much @nicolas-f @pierromond !

**Comment by nicolas-f** [comment]
Good, I think building table can contain Polygon and now LineString, so you can simulate screens. You can also provided a different absorption value for each screen with a G column field (0-1 for same coefficient on each frequency bands, or more elaborate with [wall impedance](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/blob/master/noisemodelling-pathfinder/src/main/java/org/noise_planet/noisemodelling/pathfinder/utils/AlphaUtils.java) 20-20000 )

**Comment by pierromond** [comment]
Nice ! What you are calling expected is CadnaA ?

Can you export some rays, just to check that everything is ok ?

By the way, congrats, you are becoming an advanced user of NM ;)

**Comment by IsotoCedex** [comment]
Thank you!!! I have a lot to learn and understand.
I´m testing several softwares. None in particular.
Here you are bridge rays.
[bridge_rays.zip](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/files/11837308/bridge_rays.zip)
It is dificult to me. Too many rays. On Google Earth it seems to come from below of the bridge. But on my 3D render, I can see that rays came from the top of the bridge. Some of them seem to cross trough the bridge. It may be because of the rendering.
**What do you think?**
- View from below on my render:
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/fedc7940-487e-428e-97d6-38bba084dc24)
- View from above on my render:
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/343f9f76-f38e-4994-9e1b-82435e615665)
- Google Earth with 3D Buildings
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/136c81ea-0199-4714-b432-f9f70b3d1747)

**Comment by IsotoCedex** [comment]
Hi again,
I have made a new model. Only the bridge, and with 3D receivers, in order to see where the rays come from.
Definitely there is a problem. I don´t know why, but rays make strange things :).
- Modeled area with receivers 4,8,12 and 16 m over terrain:
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/f29f955e-bd1c-463c-a26b-ba57544eb11a)
- Results on recceivers:
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/9a14a88d-0c34-402b-bd06-91cbfc1ae844)
- Rays:
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/fe3f1197-0b7f-4256-8e4d-f1fe7c2873c1)
My datasource and txt process
[07_Bridge3D.zip](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/files/11846721/07_Bridge3D.zip)

Could you please, @nicolas-f @pierromond , have a look to that?

**Comment by IsotoCedex** [comment]
Hi again @nicolas-f and @pierromond.

In the last comment, I have detected an error. The first point of the road was outside the DEM.

I have prepared two more data sets to check where the problem is.
You can download them:

[Sourcedata2.zip](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/files/11868382/Sourcedata2.zip)

Those data sets are made up of the following layers:
- Buildings.shp: fictional buildings in the corners of the model. height=0
- Buildings_bridge.shp: The same plus the bridge, HEIGHT=1 m above ground
- DEM: 2x2 DEM points, with Z data
- Road: road section of the bridge.

Thus, you can find two situations:
- With Bridge:
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/c012318b-4a2f-4075-9735-6812c4d61623)

- Without bridge:
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/2ef7aed7-0865-4349-8642-12d1cec36370)



I have calculated a noise map for these two locations. The results are the following:

- **Without the bridge** (results files: [NoBridge.zip](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/files/11868555/NoBridge.zip)

- Rays seems to be ok. No rays under the ground.
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/b5e28d8a-6594-4111-bf31-3647bef2dfde)
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/9cb106c4-617e-438e-9977-1db24115b271)

- But isophones are like if the road where directly over the terrain. There is not a screen efect between the road and the terrain:
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/95deea05-cb2e-46ab-aee6-98a9f248f485)


**___________________________________________________________________________________**

- **With the bridge** (results files: [Bridge.zip](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/files/11868772/Bridge.zip))

- Rays seems are wrong. part of them are under terrain.
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/bef0b404-d8fc-4c49-879b-b8f4205aa405)

- Isophones are not realistic.
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/f1b1320d-9b29-4fc1-9efd-ce91e52c4ff2)


My biggest concern is to know why there are rays underground if I put the bridge, and why, if I don't, it doesn't happen.

How can we know if it is well modeled and the results are reliable?

Thank you very much.

**Reply by nicolas-f** [reply]
Yes line source that are both in and out of the dem was raising an issue, it has been fixed with the development version and added to unit test.


> If you filter to keep only one sound source line you could use export_scene parameter to export kml files that contain propagation rays. (ex: file:///home/fortin/data/debug_nm/rays/map.kml) and open it in chrome at this url https://earth.google.com/

https://github.com/Universite-Gustave-Eiffel/NoiseModelling/pull/496

The file kml export as explained earlier contain all inputs elements and are using absolute altitude so it is a good way to see what final 3d model has been used. You should be able to see the modified DEM bridge you have designed.

**Comment by IsotoCedex** [comment]
Hi @nicolas-f ,
I have created several folders and tried to export kml scene. It does not work:
- Folder: **file:///C:/dir/map.kml**
- Error: 
Process failed during execution javax.script.ScriptException: java.io.FileNotFoundException: C:\dir\file:\C:\dir\map_0_0.kml (El nombre de archivo, el nombre de directorio o la sintaxis de la etiqueta del volumen no son correctos) java.io.FileNotFoundException: C:\dir\file:\C:\dir\map_0_0.kml (El nombre de archivo, el nombre de directorio o la sintaxis de la etiqueta del volumen no son correctos) C:\dir\file:\C:\dir\map_0_0.kml (El nombre de archivo, el nombre de directorio o la sintaxis de la etiqueta del volumen no son correctos)

- Other folder: **file:///C:/NMvssfl_fra/07_Bridge3D/Result2/Bridge/KML/map.kml**
- Error:
- Process failed during execution javax.script.ScriptException: java.io.FileNotFoundException: C:\NMvssfl_fra\07_Bridge3D\Result2\Bridge\KML\file:\C:\NMvssfl_fra\07_Bridge3D\Result2\Bridge\KML\map_0_0.kml (El nombre de archivo, el nombre de directorio o la sintaxis de la etiqueta del volumen no son correctos) java.io.FileNotFoundException: C:\NMvssfl_fra\07_Bridge3D\Result2\Bridge\KML\file:\C:\NMvssfl_fra\07_Bridge3D\Result2\Bridge\KML\map_0_0.kml (El nombre de archivo, el nombre de directorio o la sintaxis de la etiqueta del volumen no son correctos) C:\NMvssfl_fra\07_Bridge3D\Result2\Bridge\KML\file:\C:\NMvssfl_fra\07_Bridge3D\Result2\Bridge\KML\map_0_0.kml (El nombre de archivo, el nombre de directorio o la sintaxis de la etiqueta del volumen no son correctos)

**Reply by nicolas-f** [reply]
Maybe there is an issue with Windows© path processing in groovy script

https://github.com/Universite-Gustave-Eiffel/NoiseModelling/blob/master/wps_scripts/src/main/groovy/org/noise_planet/noisemodelling/wps/NoiseModelling/Noise_level_from_source.groovy#L577

May be it is kmlFileNamePrepend part ? I don't have this operating system so I can not fix the script.

**Reply by IsotoCedex** [reply]
>

**Reply by IsotoCedex** [reply]
Hi again @nicolas-f,
Yes, there is a problem in https://github.com/Universite-Gustave-Eiffel/NoiseModelling/blob/master/wps_scripts/src/main/groovy/org/noise_planet/noisemodelling/wps/NoiseModelling/Noise_level_from_source.groovy#L577

I got it to work by replacing the code with the following:

```
File folderExportKML = null;
String kmlFileNamePrepend = "";
if (input['confRaysName'] && !((input['confRaysName'] as String).isEmpty())) {
    String confRaysName = input['confRaysName'] as String;
    if (confRaysName.startsWith("file:///")) {
        ldenConfig.setExportRaysMethod(LDENConfig.ExportRaysMethods.TO_MEMORY);
        String filePath = confRaysName.substring(8); // Eliminar el prefijo "file:///"
        folderExportKML = new File(filePath).getParentFile();
        kmlFileNamePrepend = new File(filePath).getName();
        kmlFileNamePrepend = kmlFileNamePrepend.substring(0, kmlFileNamePrepend.lastIndexOf("."));
    } else {
        ldenConfig.setExportRaysMethod(LDENConfig.ExportRaysMethods.TO_RAYS_TABLE);
        ldenConfig.setRaysTable(confRaysName);
    }
    ldenConfig.setKeepAbsorption(true);
    ldenConfig.setMaximumRaysOutputCount(maximumRaysToExport);
}
```
I still don't know what's going on. If I view the kml with the "Google Earth relief" option turned on, I see this:
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/6f1f2819-4af9-4e00-b34c-722fc3ec6a2d)


If I disable it I see this:
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/18c7383d-32fb-4377-b9d0-c55b2edb19c3)

Bridge seems to be right:
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/9367dcca-8fc8-4896-834f-cea469304bd8)

Attached the KML
[map_0_0.zip](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/files/11881209/map_0_0.zip)


Please can you check the kml? Thank you!!

**Reply by nicolas-f** [reply]
Wait, what is the grey block ?
Bridge should be done as a DEM elevation not as a building.

Do you use earth.google.com or the google earth software ? I think it works only on the browser.

**Comment by IsotoCedex** [comment]
Wow!

I've been trying to make the bridge with a floating building for a week.

Ok, but if I do it with the DEM, how do I simulate what happens under the bridge?

 I use both, and the result is very similar.

![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/7828cde7-a2c2-4571-b4c7-5a2e46e22971)
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/ea8e9590-2561-458f-920b-39977798f5c5)


**Can you send me an example of a DEM with a bridge? Please. In order to know how to do. Thank you**

**Reply by nicolas-f** [reply]
You can't simulate stuff under a bridge as Cnossos propagation rays always follow a vertical convex hull path over the highest structures between receiver and source point.

If we follow the tutorial you should have something like that.

```sql
-- insert bridge as dem
DROP TABLE IF EXISTS DEM_WITH_BRIDGE;
CREATE TABLE DEM_WITH_BRIDGE(PK SERIAL, THE_GEOM GEOMETRY(geometry, 25830)) AS SELECT PK,THE_GEOM FROM DEM;
-- Remove dem points 16m from the road
DELETE FROM DEM_WITH_BRIDGE DEM WHERE EXISTS (SELECT 1 FROM ROADS WHERE ROADS.THE_GEOM && ST_EXPAND(DEM.THE_GEOM, 16, 16) AND ST_DISTANCE(ROADS.THE_GEOM, DEM.THE_GEOM) < 16);
-- Create points at 15m from the road with the same height as the road
INSERT INTO DEM_WITH_BRIDGE(THE_GEOM) SELECT ST_UPDATEZ(ST_ToMultiPoint(ST_DENSIFY(ST_BUFFER(THE_GEOM, 15),3)),ST_Z(THE_GEOM)) THE_GEOM FROM ROADS;

CALL GeoJsonWrite('/home/fortin/data/bridge/dem_with_bridge.geojson', 'DEM_WITH_BRIDGE', true);
```
[dem_with_bridge.zip](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/files/11882413/dem_with_bridge.zip)

**Reply by IsotoCedex** [reply]
Thank you! I Will check it.

**Comment by IsotoCedex** [comment]
Hello again @nicolas-f ,
Now I do understand, I think.
Bridges are modeled as an elevation on the ground. It is not possible to have a elevateds structure above the ground, with space below.

Considering your example I have modified my DEM.
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/049ac4c6-7386-47e4-9a76-e562b35b5195)
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/45ad1106-ab48-4dbf-b6bc-f107da332226)

Therefore, I understand that it is not possible to model complex situations like this:

![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/cd9348b2-a44b-4011-89f4-9339c79a272a)

Thank you!

**Reply by pierromond** [reply]
Hi @IsotoCedex,

In fact, the pathfinder in CNOSSOS can't go "under" a rigid structure (I think in CadnAA is the same).  In this case, you have two possibilities : 
1- is to modelled everything except the bridge
2- to a map of the bridge
3- combine the sound levels at the receivers points.

Another solution could be to just put a "flying road" crossing the "ground road". You will not model correctly the diffraction on the bridge edges, but at a city scale, that is not a big issue.

Do you have any info how commercial software manage this special cases ?

**Reply by IsotoCedex** [reply]
In commercial software it is possible to introduce bridges and acoustic screens as independent entities.

They are not part of the DEM.

In the case of roads noise, DEM only representes the ground. The bridge is introduced later, as an independent entity, with its characteristics.
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/4aa78977-0fdd-405e-9f59-9d67d405c6f4)

This way, we can take into account what happens under the bridge, in particular if there is another source - another road, railway, industrial source.

It can also be modelled, for example, as a reflective surface, in the case of a road pedestrian overpass.

![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/807b587a-0995-46a1-9e88-dbd42137e07c)


The images are from our CNOSSOS-EU application guide. Unfortunately we only have it in Spanish. You can find it this link: https://sicaweb.cedex.es/wp-content/uploads/2022/04/20220923_CNOSSOS_rev4.pdf

**Reply by pierromond** [reply]
This is their nice interface, but I am not sure to understand how they compute the rays. Can you export them in CadnaA ? Because of the way of CNOSSOS is written, I don't understand well how they can do...

**Reply by IsotoCedex** [reply]
I'm afraid I can't help you with that, in this moment. I am not calculating this model with CadnaA. I am comparing with results calculated by other colleagues, in various software.
I will  ask for rays coming from those calculations.

**Reply by pierromond** [reply]
Thanks !

**Comment by IsotoCedex** [comment]
Hello again @pierromond ,
Here you are an example of rays from another software:
[BridgeRays.zip](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/files/11905921/BridgeRays.zip)

In other software, the calculation is done in a similar way to NM.
The source (emitters), in this case the bridge, is divided into very close points with a relative height of 0.05 m above the bridge.
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/8b6aebf7-0d50-4c68-a15f-92abfb92ee74)

Receivers are generated, in regular mesh, with the mesh size indicated by the user.
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/8f0e14a4-b2cd-4646-b5f9-1dfc9ee4716b)

The attenuation of each emission point is calculated on each receiver.
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/8fd3c1d0-abde-4235-b894-c7eafdaf36d0)

The total noise in each receiver is calculated.
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/f04a5a08-12a4-4a78-b8fa-78c52a081f69)

The result is interpolated.
![image](https://github.com/Universite-Gustave-Eiffel/NoiseModelling/assets/87642224/672af0cd-8707-4b29-8ce2-cda99669ee23)


**The difference with NM**, as far as I know, is that:
- receivers are calculated a regular mesh,
- rays always go from an emitter 0.05 m above the source, to a receiver 4 m above the ground
- no rays are generated under the ground or above the height of the receiver.
- no rays along the bridge or surrounding it
- calculation seems cleaner. That is, rays represents a visual between the emitter and the receiver.
